// tgsnake/lib/index.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2021 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BotError = exports.Composer = exports.GramJs = exports.Updates = exports.Telegram = exports.Snake = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "Snake", { enumerable: true, get: function () { return client_1.Snake; } });
var Telegram_1 = require("./Telegram");
Object.defineProperty(exports, "Telegram", { enumerable: true, get: function () { return Telegram_1.Telegram; } });
exports.Updates = __importStar(require("./Update"));
exports.GramJs = __importStar(require("telegram"));
var Composer_1 = require("./Context/Composer");
Object.defineProperty(exports, "Composer", { enumerable: true, get: function () { return Composer_1.Composer; } });
var Error_1 = __importDefault(require("./Context/Error"));
exports.BotError = Error_1.default;
