// tgsnake/lib/Telegram/Messages/GetMessagesViews.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2021 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetMessagesViews = exports.ResultsMessagesViews = void 0;
var telegram_1 = require("telegram");
var ToBigInt_1 = require("../../Utils/ToBigInt");
var Error_1 = __importDefault(require("../../Context/Error"));
var ResultsMessagesViews = /** @class */ (function () {
    function ResultsMessagesViews(results) {
        this.date = Math.floor(Date.now() / 1000);
        if ((results === null || results === void 0 ? void 0 : results.views.length) > 0) {
            var tempViews = new Array();
            for (var i = 0; i < results.views.length; i++) {
                var msg = results.views[i];
                tempViews.push(new Views(msg));
            }
            this.views = tempViews;
        }
    }
    return ResultsMessagesViews;
}());
exports.ResultsMessagesViews = ResultsMessagesViews;
var Views = /** @class */ (function () {
    function Views(getMessagesViews) {
        if (getMessagesViews.views) {
            this.views = getMessagesViews.views;
        }
        if (getMessagesViews.forwards) {
            this.forwards = getMessagesViews.forwards;
        }
        if (getMessagesViews.replies) {
            this.replies = getMessagesViews.replies;
        }
    }
    return Views;
}());
/**
 * Get and increase the view counter of a message sent or forwarded from a channel.
 * @param snakeClient - Client
 * @param {number|string|bigint} chatId - Where the message was found.
 * @param {Array} messageId - IDs of message.
 * @param {boolean} increment - Whether to mark the message as viewed and increment the view counter
 * ```ts
 * bot.command("getMessagesViews",async (ctx)=>{
 *     let results = await ctx.telegram.getMessagesViews(ctx.chat.id,[ctx.id])
 *     console.log(results)
 * })
 * ```
 */
function GetMessagesViews(snakeClient, chatId, messageId, increment) {
    if (increment === void 0) { increment = false; }
    return __awaiter(this, void 0, void 0, function () {
        var mode, _a, id, type, peer, _b, error_1, botError;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _c.trys.push([0, 3, , 4]);
                    mode = ['debug', 'info'];
                    if (mode.includes(snakeClient.logger)) {
                        snakeClient.log("[".concat(snakeClient.connectTime, "] - [").concat(new Date().toLocaleString(), "] - Running telegram.getMessagesViews"));
                    }
                    return [4 /*yield*/, (0, ToBigInt_1.toBigInt)(chatId, snakeClient)];
                case 1:
                    _a = __read.apply(void 0, [_c.sent(), 3]), id = _a[0], type = _a[1], peer = _a[2];
                    _b = ResultsMessagesViews.bind;
                    return [4 /*yield*/, snakeClient.client.invoke(new telegram_1.Api.messages.GetMessagesViews({
                            peer: peer,
                            id: messageId,
                            increment: increment,
                        }))];
                case 2: return [2 /*return*/, new (_b.apply(ResultsMessagesViews, [void 0, _c.sent()]))()];
                case 3:
                    error_1 = _c.sent();
                    botError = new Error_1.default();
                    botError.error = error_1;
                    botError.functionName = 'telegram.getMessagesViews';
                    botError.functionArgs = "".concat(chatId, ",").concat(JSON.stringify(messageId), ",").concat(increment);
                    throw botError;
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.GetMessagesViews = GetMessagesViews;
