// tgsnake/lib/Telegram/Media/SendSticker.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2021 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendSticker = void 0;
var telegram_1 = require("telegram");
var SendMedia_1 = require("./SendMedia");
var UploadFile_1 = require("./UploadFile");
var tg_file_id_1 = require("tg-file-id");
var big_integer_1 = __importDefault(require("big-integer"));
var Error_1 = __importDefault(require("../../Context/Error"));
/**
 * Sending sticker with fileId/file location/url/buffer.
 * @param snakeClient - client
 * @param {bigint|number|string} chatId - Chat/Groups/Channel id.
 * @param {string|Buffer|Object} fileId - Path file/FileId/Buffer.
 * ```ts
 * bot.on("message",async (ctx) => {
 *     if(ctx.media && ctx.media.type == "sticker"){
 *         let results = await ctx.telegram.sendSticker(ctx.chat.id,ctx.media.fileId)
 *         console.log(results)
 *     }
 * })
 * ```
 */
function SendSticker(snakeClient, chatId, fileId) {
    return __awaiter(this, void 0, void 0, function () {
        var mode, final, file, decode, resultsSendSticker, accessHash, id, e_1, error_1, botError;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 12, , 13]);
                    mode = ['debug', 'info'];
                    if (mode.includes(snakeClient.logger)) {
                        snakeClient.log("[".concat(snakeClient.connectTime, "] - [").concat(new Date().toLocaleString(), "] - Running telegram.sendSticker"));
                    }
                    final = void 0;
                    if (fileId instanceof telegram_1.Api.MessageMediaDocument) {
                        final = fileId;
                    }
                    if (fileId instanceof telegram_1.Api.Document) {
                        final = fileId;
                    }
                    if (!(Buffer.isBuffer(fileId) ||
                        /^http/i.exec(String(fileId)) ||
                        /^(\/|\.\.?\/|~\/)/i.exec(String(fileId)))) return [3 /*break*/, 2];
                    return [4 /*yield*/, (0, UploadFile_1.UploadFile)(snakeClient, fileId)];
                case 1:
                    file = _a.sent();
                    final = new telegram_1.Api.InputMediaUploadedDocument({
                        file: file,
                        mimeType: 'image/webp',
                        attributes: [
                            new telegram_1.Api.DocumentAttributeFilename({
                                fileName: "sticker.webp",
                            }),
                        ],
                    });
                    _a.label = 2;
                case 2:
                    if (!final) return [3 /*break*/, 3];
                    return [2 /*return*/, (0, SendMedia_1.SendMedia)(snakeClient, chatId, final)];
                case 3:
                    decode = void 0;
                    if (typeof fileId !== 'string') {
                        throw new Error('Invalid FileId!');
                    }
                    try {
                        decode = (0, tg_file_id_1.decodeFileId)(String(fileId));
                    }
                    catch (error) {
                        throw new Error('Invalid FileId!');
                    }
                    if (!decode) return [3 /*break*/, 11];
                    if (!(decode.fileType == 'sticker')) return [3 /*break*/, 10];
                    resultsSendSticker = void 0;
                    accessHash = String(decode.access_hash);
                    id = String(decode.id);
                    _a.label = 4;
                case 4:
                    if (!true) return [3 /*break*/, 9];
                    _a.label = 5;
                case 5:
                    _a.trys.push([5, 7, , 8]);
                    return [4 /*yield*/, (0, SendMedia_1.SendMedia)(snakeClient, chatId, new telegram_1.Api.InputMediaDocument({
                            id: new telegram_1.Api.InputDocument({
                                id: (0, big_integer_1.default)(id),
                                accessHash: (0, big_integer_1.default)(accessHash),
                                fileReference: Buffer.from(decode.fileReference, 'hex'),
                            }),
                        }))];
                case 6:
                    resultsSendSticker = _a.sent();
                    return [3 /*break*/, 9];
                case 7:
                    e_1 = _a.sent();
                    if (BigInt(accessHash) > BigInt(0) && BigInt(id) > BigInt(0)) {
                        // id (+) accessHash (+)
                        accessHash = "-".concat(accessHash); // id (+) accessHash (-)
                    }
                    else if (BigInt(accessHash) < BigInt(0) && BigInt(id) > BigInt(0)) {
                        // id (+) accessHash (-)
                        id = "-".concat(id); // id (-) accessHash (-)
                    }
                    else if (BigInt(accessHash) < BigInt(0) && BigInt(id) < BigInt(0)) {
                        // id (-) accessHash (-)
                        accessHash = accessHash.replace(/^\-/, '');
                        // id (-) accessHash (+)
                    }
                    else {
                        //@ts-ignore
                        throw new Error(e_1.message);
                        return [3 /*break*/, 9];
                    }
                    return [3 /*break*/, 8];
                case 8: return [3 /*break*/, 4];
                case 9: return [2 /*return*/, resultsSendSticker];
                case 10: throw new Error("Invalid FileType. It must be \"sticker\". Received ".concat(decode.fileType));
                case 11: return [3 /*break*/, 13];
                case 12:
                    error_1 = _a.sent();
                    botError = new Error_1.default();
                    botError.error = error_1;
                    botError.functionName = 'telegram.sendSticker';
                    botError.functionArgs = "".concat(chatId, ",").concat(Buffer.isBuffer(fileId) ? "<Buffer ".concat(fileId.toString('hex'), ">") : JSON.stringify(fileId));
                    throw botError;
                case 13: return [2 /*return*/];
            }
        });
    });
}
exports.SendSticker = SendSticker;
