// tgsnake/lib/Utils/AdminRights.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2021 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminRights = void 0;
var AdminRights = /** @class */ (function () {
    function AdminRights(adminRights) {
        this.changeInfo = adminRights.changeInfo;
        this.postMessages = adminRights.postMessages;
        this.editMessages = adminRights.editMessages;
        this.deleteMessages = adminRights.deleteMessages;
        this.banUsers = adminRights.banUsers;
        this.inviteUsers = adminRights.inviteUsers;
        this.pinMessages = adminRights.pinMessages;
        this.addAdmins = adminRights.addAdmins;
        this.anonymous = adminRights.anonymous;
        this.manageCall = adminRights.manageCall;
        this.other = adminRights.other;
    }
    return AdminRights;
}());
exports.AdminRights = AdminRights;
