// tgsnake/lib/Utils/ChatPhoto.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatPhoto = void 0;
var media = __importStar(require("./Media"));
var ToBigInt_1 = require("./ToBigInt");
var tg_file_id_1 = require("tg-file-id");
var ChatPhoto = /** @class */ (function () {
    function ChatPhoto(photo, resultsGetEntity) {
        this.isBig = true;
        var file = media.generateFileId({
            fileType: 'profile_photo',
            typeId: media.typeId.CHAT_PHOTO,
            version: 4,
            subVersion: 30,
            dcId: photo.dcId,
            id: BigInt((0, ToBigInt_1.toString)(photo.photoId)),
            accessHash: BigInt(0),
            photoSizeSource: 'dialogPhoto',
            dialogId: resultsGetEntity.id,
            isSmallDialogPhoto: false,
            photoSizeSourceId: media.thumbTypeId.CHAT_PHOTO_BIG,
            dialogAccessHash: resultsGetEntity.accessHash,
            volumeId: BigInt(1),
        });
        this.fileId = file.fileId;
        this.uniqueFileId = file.uniqueFileId;
    }
    ChatPhoto.prototype.decode = function (fileId) {
        var file = fileId || this.fileId;
        if (!file)
            return "FileId not found!";
        return (0, tg_file_id_1.decodeFileId)(String(file));
    };
    return ChatPhoto;
}());
exports.ChatPhoto = ChatPhoto;
